README.md
A basic, explicit CSV parsing and formatting library for use in node.js. CSV strings are parsed and formatted using a set of column definitions for converting to and from CSV.

## Installation

    npm install salsacsv

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

#### Table of Contents

-   [salsacsv](#salsacsv)
    -   [toCSV](#tocsv)
        -   [Parameters](#parameters)
        -   [Examples](#examples)
    -   [fromCSV](#fromcsv)
        -   [Parameters](#parameters-1)
        -   [Examples](#examples-1)
    -   [cellLabel](#celllabel)
        -   [Parameters](#parameters-2)
        -   [Examples](#examples-2)
-   [Column](#column)
    -   [Properties](#properties)
-   [Converter](#converter)
    -   [Parameters](#parameters-3)
-   [ConverterDetails](#converterdetails)
    -   [Properties](#properties-1)
-   [Parser](#parser)
    -   [Parameters](#parameters-4)
-   [ParserDetails](#parserdetails)
    -   [Properties](#properties-2)

### salsacsv

Used for converting data to and from CSV.

#### toCSV

Converts an array of objects to a CSV string.

##### Parameters

-   `rows` **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)>** Array of objects to form rows from.
-   `columns` **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[Column](#column)>** An array containing columns.
-   `options` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** Parsing options. (optional, default `{}`)
    -   `options.includeHeader` **[Boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?** Whether the CSV has a header or not, the first line will be skipped if this is set to true.
    -   `options.delimiter` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The delimiter for the CSV string. (optional, default `','`)

##### Examples

```javascript
toCSV([
    {
        name: 'Cat Chow',
        price: 529
    }
], [
    {
        header: 'Name',
        key: 'name'
    },
    {
        header: 'Price',
        key: 'price',
        converter: (value) => value / 100
    }
], {
    includeHeader: true
});
// "Name","Price"\n"Cat Chow",5.29
```

Returns **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** CSV string.

#### fromCSV

Converts a CSV string into objects.

##### Parameters

-   `csvStr` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** CSV string.
-   `columns` **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[Column](#column)>** An array containing columns. (optional, default `[]`)
-   `options` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** Parsing options. (optional, default `{}`)
    -   `options.includeHeader` **[Boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?** Whether the CSV has a header or not.
    -   `options.includeEmptyValues` **[Boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?** Whether to assign empty values to object or not.
    -   `options.delimiter` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The delimiter of the CSV string. (optional, default `','`)

##### Examples

```javascript
fromCSV('"Name","Price"\n"Cat Chow",5.29', [
    {
        header: 'Name',
        key: 'name'
    },
    {
        header: 'Price',
        key: 'price',
        parser: (value) => Math.round(parseFloat(value) * 100)
    }
], {
    includeHeader: true
});
// { name: 'Cat Chow', price: 529 }
```

```javascript
// using no column definitions
// there will be no type conversions if no parser is given, any returned values will be strings
fromCSV('"Cat Chow",5.29');
// { col1: 'Cat Chow', col2: '5.29' }
```

```javascript
// using no column definitions (use header as keys)
fromCSV('"Name","Price"\n"Cat Chow",5.29', null, {
    includeHeader: true
});
// { Name: 'Cat Chow', Price: '5.29' }
```

Returns **[Array](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Array)&lt;[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)>** Array of objects.

#### cellLabel

Gets the cell label.

##### Parameters

-   `rowNumber` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Zero-based row number.
-   `columnNumber` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Zero-based column number.

##### Examples

```javascript
cellLabel(0, 3); // A3
```

Returns **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Cell label.

### Column

An object describing the format of a column.

Type: [Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)

#### Properties

-   `header` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)?** The header to be used for this column.
-   `key` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)?** The object key for this column.
-   `required` **[Boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?** Indicates whether the value should be defined when getting value from CSV. Throws an error if the resulting value is null, undefined, or an empty string.
-   `converter` **[Converter](#converter)?** The function called to convert value to CSV.
-   `parser` **[Parser](#parser)?** The function called to parse value from CSV. Will not be called unless parseEmpty is set to true.
-   `parseEmpty` **[Boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)?** Whether to parse empty values or not.

### Converter

Function to convert value to raw CSV.

Type: [Function](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/function)

#### Parameters

-   `value` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)?** Value from object.
-   `details` **[ConverterDetails](#converterdetails)?** Details of cell.

Returns **([String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String) \| [Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number) | null | [undefined](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/undefined))** Value of cell.

### ConverterDetails

Converter details

Type: [Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)

#### Properties

-   `obj` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)?** Source object.
-   `key` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)?** The key of the value we want to take from the object for this column.
-   `row` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)?** Row number.
-   `column` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)?** Column number.

### Parser

Function to parse value from raw CSV.

Type: [Function](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/function)

#### Parameters

-   `value` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)?** Value of cell.
-   `details` **[ParserDetails](#parserdetails)?** Details of cell.

Returns **any** Parsed value from CSV string.

### ParserDetails

Parser details

Type: [Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)

#### Properties

-   `key` **[String](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)?** Name of key to assign to object.
-   `row` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)?** Row number.
-   `column` **[Number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)?** Column number.

## License

MIT


CHANGELOG.md
# Changelog

## [1.0.1] - 2019-02-20
### Added
- Typescript types, better documentation.

tests/data/orders.js
'use strict';

// this is used for comprehensive testing

const {cellLabel} = require('../..');

// helper functions for our data

// converts a string to a date
const toDate = (str) => new Date(str);
// converts a date to a string to be inserted into CSV
const toDateString = (date) => {
    return [
        date.getFullYear(),
        date.getMonth() + 1,
        date.getDate()
    ].join('/');
};
// this will convert our currency value into an integer
// e.g. 1.23 -> 123
const toIntegerCurrency = (str) => {
    return Math.round(parseFloat(str) * 100);
};
// reverse of toIntegerCurrency
// this will convert our integer price value into a decimal number
// e.g. 123 -> 1.23
const toDecimalCurrency = (num) => {
    const toDecimal = (value, precision) => {
        let power = Math.pow(10, precision);
        
        return value / power;
    };
    
    return toDecimal(num, 2);
};
// don't forget the sales tax!
const calculateTax = (tax, price) => {
    return toDecimalCurrency(Math.round(price * tax));
};

const tax = 0.06;
// our columns for our order data
const columns = [
    {
        header: 'Date',
        key: 'date',
        required: true,
        // when parsing from CSV the result should be the same as the original object
        parser: toDate,
        converter: toDateString
    },
    // no parsing needed
    {
        header: 'Name',
        key: 'name',
        required: true
    },
    {
        header: 'Price',
        key: 'price',
        required: true,
        // when parsing from CSV the result should be the same as the original object
        parser: toIntegerCurrency,
        converter: toDecimalCurrency
    },
    // this is a special column for CSV only to show the tax added to an item
    {
        header: 'Tax',
        converter: (value, {obj}) => {
            return calculateTax(tax, obj.price);
        }
    },
    // this is a special column for CSV-only to show the total of an item with tax
    {
        header: 'Total',
        // this generates a spreadsheet formula to take the price cell and the tax cell
        // and add them together
        converter: (value, {row, column}) => {
            let priceCell = cellLabel(row, column - 2);
            let taxCell = cellLabel(row, column - 1);
            
            return `=${priceCell}+${taxCell}`;
        }
    }
];
// our columns for our orders
const data = [
    {
        date: new Date(2019, 7, 25),
        name: 'Cat Chow',
        price: 349
    },
    {
        date: new Date(2019, 7, 26),
        name: 'Water',
        price: 129
    },
    {
        date: new Date(2019, 7, 26),
        name: 'Light Bulbs',
        price: 529
    }
];

module.exports = {
    data,
    columns
};

tests/helpers.test.js
'use strict';

// test our helper functions

const {cellLabel} = require('..');

it('Gets correct cell label (row 10, column 5)', () => {
    // rows and columns are 0-indexed
    const row = 10 - 1;
    const column = 5 - 1;
    const label = cellLabel(row, column);
    
    expect(label).toBe('E9');
});

tests/conversions.test.js
'use strict';

// tests for converting from/to CSV

const {toCSV, fromCSV} = require('..');
const ordersData = require('./data/orders');

it('Converts to CSV correctly', () => {
    const columns = [
        'name',
        {
            key: 'amount',
            parser: Number
        }
    ];
    const data = [
        {
            name: 'Kitty cat',
            amount: 2
        }
    ];
    const csv = toCSV(data, columns);
    const expected = '"Kitty cat",2';
    
    expect(csv)
        .toBe(expected);
});

it('Converts from CSV correctly', () => {
    const columns = [
        'name',
        {
            key: 'amount',
            parser: Number
        }
    ];
    const csv =  '"Kitty cat",2';
    const parsed = fromCSV(csv, columns);
    
    expect(parsed)
        .toEqual([
            {
                name: 'Kitty cat',
                amount: 2
            }
        ]);
});

it('Converts to CSV correctly (comprehensive)', () => {
    const {columns, data} = ordersData;
    // we will test the first order in the array
    const testOrders = [data[0]];
    // expected result of the first order
    const expected = '2019/8/25,"Cat Chow",3.49,0.21,=C1+D1';
    const converted = toCSV(testOrders, columns);
    
    expect(converted)
        .toBe(expected);
});

it('Parses from CSV correctly (comprehensive)', () => {
    const {columns, data} = ordersData;
    // we will test the first order in the array
    const testOrders = [data[0]];
    const testCSV = '2019/8/25,"Cat Chow",3.49,0.21,=C1+D1';
    const parsed = fromCSV(testCSV, columns);
    
    expect(parsed)
        .toEqual(testOrders);
});

// convert an array of objects to CSV
// then convert the CSV back to objects
// and check if they are equal
it('Converts to and from CSV correctly (comprehensive)', () => {
    const {columns, data} = ordersData;
    const testOrders = data;
    const csv = toCSV(testOrders, columns);
    const rows = fromCSV(csv, columns);
    
    expect(rows)
        .toEqual(testOrders);
});

it('Ignores empty lines', () => {
    const csv = '"Cat Chow",5.29\n\n\n\n\n"Pizza",6.99\n\n\n"Bowl",2.69';
    const parsed = fromCSV(csv);
    
    expect(parsed.length)
        .toBe(3);
});

it('Assigns header columns as key names', () => {
    const csv = '"Name","Price"\n"Cat Chow",5.29\n"Pizza",6.99';
    const parsed = fromCSV(csv, null, {
        includeHeader: true
    });
    // this will maintain the same order as their appearance
    const keys = Object.keys(parsed[0]);
    
    expect(keys)
        .toEqual([
            'Name',
            'Price'
        ]);
});

it('Properly converts objects to CSV without columns input', () => {
    const data = [
        {
            name: 'Cat Chow',
            price: 529
        },
        {
            name: 'Pizza',
            price: 699
        }
    ];
    const csv = toCSV(data, null, {
        includeHeader: true
    });
    
    expect(csv)
        .toBe('"name","price"\n"Cat Chow",529\n"Pizza",699');
});

it('Ignores object values', () => {
    const data = [
        {
            name: 'Cat Chow',
            price: 529,
            data: {
                sku: '53;3'
            }
        },
        {
            name: 'Pizza',
            price: 699,
            data: {
                sku: '109;2'
            }
        }
    ];
    const csv = toCSV(data, null, {
        includeHeader: true
    });
    
    expect(csv)
        .toEqual('"name","price","data"\n"Cat Chow",529,\n"Pizza",699,');
});

index.js
/**
 * Escapes a string in CSV.
 * @param {String} str - String.
 * @returns {String} Escaped string.
 * @private
 */
function escapeCSV(str) {
    // quotes are replaced with double quotes to escape them
    // https://www.freeformatter.com/csv-escape.html
    return `"${str.toString().replace(/"/g, '""')}"`;
}

/**
 * Helper function to detect a column.
 * @param {*} column - Column to check.
 * @returns {(Column|null)} Column object, or null if column is not valid.
 * @private
 */
function detectColumn(column) {
    const columnIsObject = Boolean(
        typeof column === 'object' &&
        // since typeof null is an object
        column !== null
    );
    const columnIsString = Boolean(
        typeof column === 'string'
    );
    
    // the column is valid
    if (columnIsString && column.length > 0) {
        // convert column to an object so that we can use it in the same manner
        return {
            key: column
        };
    } else if (columnIsObject) {
        // it is an object and it has a key
        return column;
    }
    
    return null;
}

/**
 * Detects columns from a CSV string.
 * @param {String} csvStr - CSV string.
 * @param {Object} options - Options.
 * @param {Boolean} [options.includeHeader] - Whether the CSV has a header or not.
 * @param {String} [options.delimiter=','] - The delimiter of the CSV string.
 * @returns {Column[]} Array of columns.
 * @private
 */
function detectColumnsFromCSV(csvStr, options = {}) {
    const {includeHeader} = options;
    const delimiter = options.delimiter || ',';
    const firstLine = csvStr.split('\n')[0];
    // get first line
    const split = csvToArray(firstLine, delimiter)[0] || [];
    let columns;
    
    // first line is header values
    if (includeHeader) {
        columns = split;
    } else {
        // setup some column names
        columns = split.map((value, i) => {
            return `col${(i + 1)}`;
        });
    }
    
    return columns.map(detectColumn);
}

/**
 * Detects columns from an array of objects.
 * @param {Object[]} rows - Array of objects to take columns from.
 * @param {Object} options - Options.
 * @param {Boolean} [options.includeHeader] - Whether the CSV has a header or not.
 * @returns {Column[]} Array of columns.
 * @private
 */
function detectColumnsFromJSON(rows, options) {
    const {includeHeader} = options;
    const keys = Object.keys(rows[0] || {});
    const format = (column, i) => {
        const canIncludeHeader = Boolean(
            column !== null &&
            includeHeader
        );
        
        if (canIncludeHeader) {
            const key = keys[i];
            
            // include the name of the key as the header for this column
            column.header = key;
        }
        
        return column;
    };
    
    return keys .map(detectColumn).map(format);
}

/**
 * Gets the column label of a column given its number index.
 * @param {Number} columnNumber - Zero-based column number.
 * @returns {String} Column letter.
 * @private
 *
 * @example
 * cellLabel(2); // C
 */
function columnLabel(columnNumber) {
    const letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    const letter = letters.charAt(columnNumber % 26);
    const concat = Math.floor(columnNumber / 26);
    
    if (concat > 0) {
        return columnLabel(concat - 1) + letter;
    }
    
    return letter;
}

/**
 * Converts a CSV string into an array of arrays where each line is split by its delimiter.
 *
 * @param {String} str - String to parse.
 * @param {String} [delimiter=','] - Delimiter.
 * @returns {Array[]} Array of arrays for each line in string.
 * @private
 */
function csvToArray(str, delimiter = ',') {
    // Source:
    // https://www.bennadel.com/blog/1504-ask-ben-parsing-csv-strings-with-javascript-exec-regular-expression-command.htm
    // https://gist.github.com/bennadel/9753411#file-code-1-htm 
    // Create a regular expression to parse the CSV values
    const objPattern = new RegExp(
        (
            // delimiters
            `(\\${delimiter}|\\r?\\n|\\r|^)` +
            // quoted fields
            '(?:"([^"]*(?:""[^"]*)*)"|' +
            // standard fields (numbers, dates, unqouted text fields)
            `([^"\\${delimiter}\\r\\n]*))`
        ),
        'gi'
    );
    
    // Create an array to hold our data. Give the array
    // a default empty first row
    let results = [[]];
    // Create an array to hold our individual pattern
    // matching groups
    let matches = null;
    
    // Keep looping over the regular expression matches
    // until we can no longer find a match
    while (matches = objPattern.exec(str)) {
        // Get the delimiter that was found
        const matchedDelimiter = matches[1];
        const hasRowDelimiter = Boolean(
            matchedDelimiter.length > 0 &&
            matchedDelimiter !== delimiter
        );
        let value;
        
        // Check to see if the given delimiter has a length
        // (is not the start of string) and if it matches
        // field delimiter. If id does not, then we know
        // that this delimiter is a row delimiter
        if (hasRowDelimiter) {
            // Since we have reached a new row of data,
            // add an empty row to our data array
            results.push([]);
        }
        
        // Now that we have our delimiter out of the way,
        // let's check to see which kind of value we
        // captured (quoted or unquoted)
        if (matches[2]) {
            // We found a quoted value. When we capture
            // this value, unescape any double quotes
            value = matches[2].replace(/""/g, '"');
        } else {
            // we found a non-quoted value
            value = matches[3];
        }
        
        // Now that we have our value string, let's add
        // it to the data array
        results[results.length - 1].push(value);
    }
    
    // Return the parsed data
    return results;
}

/**
 * Converts an array of objects to a CSV string.
 * @param {Object[]} rows - Array of objects to form rows from.
 * @param {Column[]} columns - An array containing columns.
 * @param {Object} [options={}] - Parsing options.
 * @param {Boolean} [options.includeHeader] - Whether the CSV has a header or not, the first line will be skipped if this is set to true.
 * @param {String} [options.delimiter=','] - The delimiter for the CSV string.
 * @returns {String} CSV string.
 * @public
 * @memberof salsacsv
 *
 * @example
 * toCSV([
 *     {
 *         name: 'Cat Chow',
 *         price: 529
 *     }
 * ], [
 *     {
 *         header: 'Name',
 *         key: 'name'
 *     },
 *     {
 *         header: 'Price',
 *         key: 'price',
 *         converter: (value) => value / 100
 *     }
 * ], {
 *     includeHeader: true
 * });
 * // "Name","Price"\n"Cat Chow",5.29
 */
function toCSV(rows, columns, options = {}) {
    if (!Array.isArray(rows)) {
        throw new Error('First argument is not an array');
    }
    
    if (!columns) {
        columns = detectColumnsFromJSON(rows, options);
    } else {
        // firstly, format columns into the proper format
        columns = columns.map(detectColumn);
    }
    
    const {includeHeader} = options;
    const delimiter = options.delimiter || ',';
    const startIndex = includeHeader ? 1 : 0;
    const getLine = (obj, rowIndex) => {
        const processColumn = (column, columnIndex) => {
            // the column is invalid
            if (column == null) {
                // empty cell
                return '';
            }
            
            const {key, converter} = column;
            let cellValue = obj[key];
            
            if (typeof converter === 'function') {
                // convert the cell value
                cellValue = converter(cellValue, {
                    obj,
                    key,
                    row: rowIndex + startIndex + 1,
                    column: columnIndex
                });
            }
            
            const canClear = Boolean(
                // cell value is null or undefined
                cellValue == null ||
                // cell value is an object
                typeof cellValue === 'object'
            );
            
            // we do not want to store these values to plain text
            if (canClear) {
                // make it an empty string instead
                cellValue = '';
            }
            
            // whether the cell value should be escaped or not
            const shouldEscape = Boolean(
                // strings shoyld be escaped
                typeof cellValue === 'string' &&
                // but only if the string is not empty
                cellValue.length > 0 &&
                // and is not a formula
                !/^=/.test(cellValue) &&
                // and also does not look like a date
                // TODO probably use a more comprehensive date pattern test
                !/^\d+[-\/]\d+[-\/]\d+$/.test(cellValue)
            );
            
            if (shouldEscape) {
                // escape strings
                cellValue = escapeCSV(cellValue);
            }
            
            return cellValue;
        };
        // collect line from columns
        const result = columns
            .map(processColumn)
            .join(delimiter);
        
        return result;
    };
    // build our lines
    let lines = rows.map(getLine);
    
    // we want to include the header
    if (includeHeader) {
        // build the header
        const header = columns.map((column) => {
            const header = (
                column &&
                column.header
            );
            
            if (typeof header === 'string' && header.length > 0) {
                return escapeCSV(header);
            } else {
                return '';
            }
        }).join(delimiter);
        
        // include the header before the lines
        lines = [header, ...lines];
    }
    
    // tie the lines together with line breaks
    const result = lines.join('\n');
    
    // all done, don't we feel accomplished?
    // no? well, alright then...
    return result;
}


/**
 * Converts a CSV string into objects.
 * @param {String} csvStr - CSV string.
 * @param {Column[]} [columns=[]] - An array containing columns.
 * @param {Object} [options={}] - Parsing options.
 * @param {Boolean} [options.includeHeader] - Whether the CSV has a header or not.
 * @param {Boolean} [options.includeEmptyValues] - Whether to assign empty values to object or not.
 * @param {String} [options.delimiter=','] - The delimiter of the CSV string.
 * @returns {Object[]} Array of objects.
 * @public
 * @memberof salsacsv
 * 
 * @example
 * fromCSV('"Name","Price"\n"Cat Chow",5.29', [
 *     {
 *         header: 'Name',
 *         key: 'name'
 *     },
 *     {
 *         header: 'Price',
 *         key: 'price',
 *         parser: (value) => Math.round(parseFloat(value) * 100)
 *     }
 * ], {
 *     includeHeader: true
 * });
 * // { name: 'Cat Chow', price: 529 }
 * @example
 * // using no column definitions
 * // there will be no type conversions if no parser is given, any returned values will be strings
 * fromCSV('"Cat Chow",5.29');
 * // { col1: 'Cat Chow', col2: '5.29' }
 * @example
 * // using no column definitions (use header as keys)
 * fromCSV('"Name","Price"\n"Cat Chow",5.29', null, {
 *     includeHeader: true
 * });
 * // { Name: 'Cat Chow', Price: '5.29' }
 */
function fromCSV(csvStr, columns, options = {}) {
    if (typeof csvStr !== 'string') {
        throw new Error('First argument is not a string');
    }
    
    csvStr = csvStr
        // clear lines that are empty
        .replace(/\n{2,}/g, '\n')
        // remove newlines at beginning and end of file
        .replace(/(\n$|^\n)/, '');
    
    const {includeEmptyValues, includeHeader} = options;
    const delimiter = options.delimiter || ',';
    const startIndex = includeHeader ? 1 : 0;
    const useDefaultColumns = Boolean(
        // no columns given
        !columns &&
        // csv string is not empty
        csvStr.length > 0
    );
    
    if (useDefaultColumns) {
        columns = detectColumnsFromCSV(csvStr, options);
    } else {
        // firstly, format columns into the proper format
        columns = columns.map(detectColumn);
    }
    
    // this will skip the first line if includeHeader is true
    // convert the CSV string into an array of arrays for each line
    const csvLines = csvToArray(csvStr, delimiter).slice(startIndex);
    const getRow = (line, rowIndex) => {
        const processColumn = (result, column, columnIndex) => {
            const validColumn = Boolean(
                column != null &&
                column.key
            );
            
            // column is not valid for taking value from
            if (!validColumn) {
                // skip this column
                return result;
            }
            
            // get the value from the line
            const {key, required, parser, parseEmpty} = column;
            let cellValue = line[columnIndex];
            // can we parse the cell?
            const canParse = Boolean(
                // a value exists on the line
                // this includes empty values
                cellValue !== undefined &&
                // parser is a function
                typeof parser === 'function' &&
                (
                    // we allow parsing of empty values
                    parseEmpty ||
                    // value is not empty
                    cellValue !== ''
                )
            );
            
            if (canParse) {
                // parse value using parsing function
                cellValue = parser(cellValue, {
                    key,
                    row: rowIndex + startIndex + 1,
                    column: columnIndex
                });
            }
            
            // the cell value is empty
            const isEmpty = Boolean (
                // the value is null or undefined
                cellValue == null ||
                // the value is a blank string
                cellValue === ''
            );
            // determine if the value can be assigned to the object
            const canAssign = Boolean(
                // we allow empty values
                includeEmptyValues ||
                !isEmpty
            );
            // there's an error in this column...
            const hasError = (
                required &&
                isEmpty
            );
            
            if (hasError) {
                // throw the error that the value for this column is missing
                throw new Error(`Required column ${key} is empty`);
            } else if (canAssign) {
                // we don't want to assign keys with empty values unless specified in options
                result[key] = cellValue;
            }
            
            return result;
        };
        // collect object from columns
        const result = columns.reduce(processColumn, {});
        
        return result;
    };
    const rows = csvLines.map(getRow);
    
    return rows;
}

/**
 * Gets the cell label.
 * @param {Number} rowNumber - Zero-based row number.
 * @param {Number} columnNumber - Zero-based column number.
 * @returns {String} Cell label.
 * @public
 * @memberof salsacsv
 *
 * @example
 * cellLabel(0, 3); // A3
 */
function cellLabel(rowNumber, columnNumber) {
    return columnLabel(columnNumber) + rowNumber;
}

/**
 * An object describing the format of a column.
 * @typedef {Object} Column
 * @property {String} [header] - The header to be used for this column.
 * @property {String} [key] - The object key for this column.
 * @property {Boolean} [required] - Indicates whether the value should be defined when getting value from CSV. Throws an error if the resulting value is null, undefined, or an empty string.
 * @property {Converter} [converter] - The function called to convert value to CSV.
 * @property {Parser} [parser] - The function called to parse value from CSV. Will not be called unless parseEmpty is set to true.
 * @property {Boolean} [parseEmpty] - Whether to parse empty values or not.
 */

/**
 * Function to convert value to raw CSV.
 * @callback Converter
 * @param {String} [value] - Value from object.
 * @param {ConverterDetails} [details] - Details of cell.
 * @returns {(String|Number|null|undefined)} Value of cell. 
 */

/**
 * Converter details
 * @typedef {Object} ConverterDetails
 * @property {Object} [obj] - Source object.
 * @property {String} [key] - The key of the value we want to take from the object for this column.
 * @property {Number} [row] - Row number.
 * @property {Number} [column] - Column number.
 */

/**
 * Function to parse value from raw CSV.
 * @callback Parser
 * @param {String} [value] - Value of cell.
 * @param {ParserDetails} [details] - Details of cell.
 * @returns {*} Parsed value from CSV string.
 */

/**
 * Parser details
 * @typedef {Object} ParserDetails
 * @property {String} [key] - Name of key to assign to object.
 * @property {Number} [row] - Row number.
 * @property {Number} [column] - Column number.
 */

/**
 * Used for converting data to and from CSV.
 * @module salsacsv
 */
module.exports = {
    toCSV,
    fromCSV,
    cellLabel
};

examples/orders.js
'use strict';

const salsacsv = require('salsacsv');

// helper functions for our data

// converts a string to a date
const toDate = (str) => new Date(str);
// converts a date to a string to be inserted into CSV
const toDateString = (date) => {
    return [
        date.getFullYear(),
        date.getMonth() + 1,
        date.getDate()
    ].join('/');
};
// this will convert our currency value into an integer
// e.g. 1.23 -> 123
const toIntegerCurrency = (str) => {
    return Math.round(parseFloat(str) * 100);
};
// reverse of toIntegerCurrency
// this will convert our integer price value into a decimal number
// e.g. 123 -> 1.23
const toDecimalCurrency = (num) => {
    const toDecimal = (value, precision) => {
        let power = Math.pow(10, precision);
        
        return value / power;
    };
    
    return toDecimal(num, 2);
};
// don't forget the sales tax!
const calculateTax = (tax, price) => {
    return toDecimalCurrency(Math.round(price * tax));
};

const tax = 0.06;
// our columns for our order data
const columns = [
    {
        header: 'Date',
        key: 'date',
        required: true,
        // when parsing from CSV the result should be the same as the original object
        parser: toDate,
        converter: toDateString
    },
    // no parsing needed
    {
        header: 'Name',
        key: 'name',
        required: true
    },
    {
        header: 'Price',
        key: 'price',
        required: true,
        // when parsing from CSV the result should be the same as the original object
        parser: toIntegerCurrency,
        converter: toDecimalCurrency
    },
    // this is a special column for CSV only to show the tax added to an item
    {
        header: 'Tax',
        converter: (value, {obj}) => {
            return calculateTax(tax, obj.price);
        }
    },
    // this is a special column for CSV-only to show the total of an item with tax
    {
        header: 'Total',
        // this generates a spreadsheet formula to take the price cell and the tax cell
        // and add them together
        converter: (value, {row, column}) => {
            let priceCell = salsacsv.cellLabel(row, column - 2);
            let taxCell = salsacsv.cellLabel(row, column - 1);
            
            return `=${priceCell}+${taxCell}`;
        }
    }
];
// our data
const data = [
    {
        date: new Date(2019, 7, 25),
        name: 'Cat Chow',
        price: 349
    },
    {
        date: new Date(2019, 7, 26),
        name: 'Water',
        price: 129
    },
    {
        date: new Date(2019, 7, 26),
        name: 'Light Bulbs',
        price: 529
    }
];
const options = {
    includeHeader: true
};

const csv = salsacsv.toCSV(data, columns, options);

console.log(csv);
/*
"Date","Name","Price","Tax","Total"
2019/8/25,"Cat Chow",3.49,0.21,=C2+D2
2019/8/26,"Water",1.29,0.08,=C3+D3
2019/8/26,"Light Bulbs",5.29,0.32,=C4+D4
*/

const orders = salsacsv.fromCSV(csv, columns, options);

console.log(orders);
/*
[
    { date: 2019-08-25T04:00:00.000Z, name: 'Cat Chow', price: 349 },
    { date: 2019-08-26T04:00:00.000Z, name: 'Water', price: 129 },
    { date: 2019-08-26T04:00:00.000Z, name: 'Light Bulbs', price: 529 }
]
*/
